src/Compiler/Service/FSharpCheckerResults.fs

```diff
@@ -909,6 +954,50 @@ type internal TypeCheckInfo
                        | ValueNone, ValueSome y -> Some y
                        | ValueNone, ValueNone -> None
 
+    /// Get completions on an identifier in a match clause
+    let GetMatchCompletionsAtPosition (identifierRange: range, parseResultsOpt, lineStr, origLongIdentOpt, colAtEndOfNamesAndResidue,
+        residueOpt, lastDotPos, line, loc, filterCtors, resolveOverloads) =
+
+        match GetUnionCasesAndActivePatternsEnvironmentLookupResolutionsAtPosition identifierRange.End with
+        | Some (isUnionInScopeAsUnqualified, items, denv, range) ->
+            let items =
+                items
+                |> List.map (fun item ->
+                    let unresolved =
+                        match item with
+                        | Item.UnionCase (uci, requiresQualifiedAccess) when not isUnionInScopeAsUnqualified ->
+                            Some {
+                                FullName = uci.Tycon.CompiledRepresentationForNamedType.FullName
+                                Namespace = trimPathByDisplayEnvList denv uci.Tycon.CompilationPath.DemangledPath |> List.toArray
+                                DisplayName = if requiresQualifiedAccess then $"{uci.Tycon.DisplayName}.{uci.DisplayName}" else uci.DisplayName
+                            }
+                        | _ -> None
+
+                    {
+                        ItemWithInst = ItemWithNoInst item
+                        MinorPriority = 0
+                        Kind = CompletionItemKind.Other
+                        IsOwnMember = false
+                        Type = None
+                        Unresolved = unresolved
+                    })
+
+            Some (items, denv, range)
+        | _ ->
```

Comment:
Somehow this shouldn't be a separate case - we should be incorporating all the logic into one function, which independently assess the relevance of union cases, then active patterns, then exception definitions, then module/namespace names, then type names, then literals etc.

For example we can also also filter literals based on type.

Reply:
Always call `GetDeclaredItems` and then cherry pick items types based on the matching context (is it a DU, enum, etc.)? That would not work when matching a DU, which is out of scope - its cases are not returned by `GetDeclaredItems`. I suppose I could then append the out-of-scope cases from `GetUnionCasesAndActivePatternsEnvironmentLookupResolutionsAtPosition`, but the whole thing would not end up much prettier.